src/simsopt/field/mgrid.py

```diff
@@ -0,0 +1,290 @@
+        nphi: number of azimuthal points, in one field period
+        nfp: number of field periods
+        rmin: minimum r grid point
+        rmax: maximum r grid point
+        zmin: minimum z grid point
+        zmax: maximum z grid point
+
+    Note the (r,z) dimensions include both end points. The (phi) dimension includes the start point and excludes the end point.
+    '''
+
+    def __init__(self,  # fname='temp', #binary=False,
+                 nr: int = 51, 
+                 nz: int = 51, 
+                 nphi: int = 24, 
+                 nfp: int = 2,
+                 rmin: float = 0.20, 
+                 rmax: float = 0.40, 
+                 zmin: float = -0.10, 
+                 zmax: float = 0.10,
+                 ):
+
+        self.nr = nr
+        self.nz = nz
+        self.nphi = nphi
+        self.nfp = nfp
+
+        self.rmin = rmin
+        self.rmax = rmax
+        self.zmin = zmin
+        self.zmax = zmax
+
+        self.n_ext_cur = 0
+        self.coil_names = []
+
+        self.br_arr = [] 
+        self.bz_arr = [] 
+        self.bp_arr = [] 
+
+    def add_field_cylindrical(self, br, bp, bz, name=None):
+        '''
+        This function saves the vector field B.
+        B is defined by cylindrical components.
+
+        The Mgrid array assumes B is sampled linearly first in r, then z, and last phi.
+        Python arrays use the opposite convention such that B[0] gives a (r,z) square at const phi
+        and B[0,0] gives a radial line and const phi and z.
+
+        It is assumed that the (br,bp,bz) inputs for this function is already in a
+        (nphi, nz, nr) shaped array.
+
+        This function may be called once for each coil group, 
+        to save sets of fields that can be scaled using EXTCUR in VMEC.
+
+        Args:
+            br: the radial component of B field
+            bp: the azimuthal component of B field
+            bz: the axial component of B field
+            name: Name of the coil group
+        '''
+
+        # appending B field to an array for all coil groups.
+        self.br_arr.append(br)
+        self.bz_arr.append(bz)
+        self.bp_arr.append(bp)
+
+        # add coil label
+        if (name is None):
+            label = _pad_string('magnet_%i' % self.n_ext_cur)
+        else:
+            label = _pad_string(name)
+        self.coil_names.append(label)
+        self.n_ext_cur = self.n_ext_cur + 1
+
+        # TO-DO: this function could check for size consistency, between different fields, and for the (nr,nphi,nz) settings of a given instance
+
+    def write(self, filename):
+        '''
+        Export class data as a netCDF binary.
+
+        The field data is represented as a single "current group". For
+        free-boundary vmec, the "extcur" array should have a single nonzero
+        element, set to 1.0.
+
+        Args:
+            filename: output file name
+        '''
+
+        ds = netcdf_file(filename, 'w', mmap=False)
```

Comment:
Context manager is helpful in case code crashes for some reason.

Reply:
Done.